### PR TITLE
GSCHED-701: start on GppProgramProvider, atom parsing

### PR DESCRIPTION
**Pull request by bryanmiller**
Started creating GppProgramProvider with a parallel organization to the OcsProgramProvider. Atom parsing from a GPP query can be tested with the gpp_atoms.py script.

### CODE REVIEW

File: scheduler/core/programprovider/gpp/gppprogramprovider.py [hunk 1]
```diff
@@ -0,0 +1,1333 @@
...
+    """
+
+    _GPI_FILTER_WAVELENGTHS = {'Y': 1.05, 'J': 1.25, 'H': 1.65, 'K1': 2.05, 'K2': 2.25}
+    _NIFS_FILTER_WAVELENGTHS = {'ZJ': 1.05, 'JH': 1.25, 'HK': 2.20}
+    _CAL_OBSERVE_TYPES = frozenset(['FLAT', 'ARC', 'DARK', 'BIAS'])
+
+    _site_for_inst = {'GMOS_NORTH': Site.GN, 'GMOS_SOUTH': Site.GS}
+
+    # GPP GMOS built-in GPU name to barcode
+    # ToDo: Eventually this needs to come from another source, e.g. Resource, ICTD, decide whether to use name or barcode
```
Comment (cquiroz) [inline scheduler/core/programprovider/gpp/gppprogramprovider.py:106]: I guess  Resource will take all the functions of ictd eventually

File: scheduler/core/programprovider/gpp/gppprogramprovider.py [hunk 2]
```diff
@@ -0,0 +1,1333 @@
...
+        IQ = 'iq'
+        SB = 'sb'
+        WV = 'wv'
+        ELEVATION_TYPE = 'elevationConstraintType'
+        ELEVATION_MIN = 'elevationConstraintMin'
+        ELEVATION_MAX = 'elevationConstraintMax'
+        TIMING_WINDOWS = 'timingWindows'
+
+    class _AtomKeys:
+        OBS_CLASS = 'class'
```
Comment (cquiroz) [inline scheduler/core/programprovider/gpp/gppprogramprovider.py:245]: I thought we got rid of `obsClass` on gpp, maybe this is something different

File: scheduler/core/programprovider/gpp/gppprogramprovider.py [hunk 3]
```diff
@@ -0,0 +1,1333 @@
...
+                if GppProgramProvider._AtomKeys.INST_NAME in step:
+                    inst_key = GppProgramProvider._AtomKeys.INST_NAME
+                instrument = step[inst_key].split(' ')[0]
+
+                if instrument in GppProgramProvider.FPU_FOR_INSTRUMENT:
+                    if GppProgramProvider.FPU_FOR_INSTRUMENT[instrument] in step:
+                        fpu = step[GppProgramProvider.FPU_FOR_INSTRUMENT[instrument]]
+
+                if instrument == 'GMOS-N' and fpu == 'IFU Left Slit (blue)':
+                    instrument = 'GRACES'
```
Comment (cquiroz) [inline scheduler/core/programprovider/gpp/gppprogramprovider.py:604]: This seems a bit odd, how does it become `GRACES`?

File: scheduler/core/programprovider/gpp/gppprogramprovider.py [hunk 4]
```diff
@@ -0,0 +1,1333 @@
...
+        scheduling_group_keys = sorted(key for key in data
+                                       if key.startswith(GppProgramProvider._GroupKeys.SCHEDULING_GROUP))
+        for key in scheduling_group_keys:
+            subgroup_id = GroupID(key.split('-')[-1])
+            subgroup = self.parse_and_group(data[key], program_id, subgroup_id, split=split,
+                                            split_by_iterator=split_by_iterator)
+            if subgroup is not None:
+                children.append(subgroup)
+
+        # Grab the observation data from any organizational folders.
```
Comment (cquiroz) [inline scheduler/core/programprovider/gpp/gppprogramprovider.py:1102]: Do we need to represent org folders in the ODB?

**Review by cquiroz — APPROVED** [review]
LGTM, looks fine, I added a couple of small questions inline

File: scheduler/core/programprovider/gpp/gppprogramprovider.py (hunk 1 above)
Comment (bryanmiller) [reply scheduler/core/programprovider/gpp/gppprogramprovider.py:106]: yes

File: scheduler/core/programprovider/gpp/gppprogramprovider.py (hunk 3 above)
Comment (bryanmiller) [reply scheduler/core/programprovider/gpp/gppprogramprovider.py:604]: This is left over from the OCS. We won't need to support this in GPP.

File: scheduler/core/programprovider/gpp/gppprogramprovider.py (hunk 4 above)
Comment (bryanmiller) [reply scheduler/core/programprovider/gpp/gppprogramprovider.py:1102]: Not at the moment.

**Pull request merged**